File: src/analysis.py
from src.park_model import calculate_drag_loss
from src.physics import score_setup_efficiency, analyze_setup_safety
from src.models import BowSetup, ArrowSetup

class VirtualCoach:
    def __init__(self, bow: BowSetup, arrow: ArrowSetup):
        self.bow = bow
        self.arrow = arrow
        
    def analyze_session_performance(self, 
                                  short_score: float, short_dist: float, short_face: int,
                                  long_score: float, long_dist: float, long_face: int):
        """
        Full analysis pipeline:
        1. Check Equipment Safety
        2. Check Setup Efficiency for the context
        3. Calculate Drag/Skill Loss
        """
        
        # 1. Safety
        safety_warnings = analyze_setup_safety(self.bow, self.arrow)
        
        # 2. Efficiency (Assume Outdoor context if long_dist > 30)
        discipline = 'outdoor' if long_dist > 30 else 'indoor'
        setup_analysis = score_setup_efficiency(self.bow, self.arrow, discipline)
        
        # 3. Park Model Analysis
        drag_analysis = calculate_drag_loss(
            short_score, short_dist, short_face,
            long_score, long_dist, long_face
        )
        
        # 4. Synthesis
        recommendations = []
        if drag_analysis['percent_loss'] > 10.0:
            recommendations.append("High Drag Loss detected (>10%).")
            if setup_analysis['gpp'] > 9.0:
                recommendations.append("-> Your arrow is heavy (High GPP). Consider lighter points for 50m.")
            if self.arrow.shaft_diameter_mm > 6.0:
                recommendations.append("-> Your arrow is thick. Wind drift is likely the cause.")
            if self.bow.tiller.type == 'positive':
                recommendations.append("-> Check Tiller. Barebow usually requires Neutral/Negative tiller.")
                
        return {
            "safety": safety_warnings,
            "setup_score": setup_analysis,
            "performance_metrics": drag_analysis,
            "coach_recommendations": recommendations
        }

File: frontend/src/domain/analysis.ts
/**
 * VirtualCoach — orchestrates physics + Park model for session analysis.
 * Ported from src/analysis.py.
 */

import type { BowSetup, ArrowSetup } from "../types/models";
import { analyzeSetupSafety, scoreSetupEfficiency } from "./physics";
import { calculateDragLoss } from "./parkModel";

export interface CoachAnalysis {
  safety: string[];
  setup_score: { score: number; gpp: number; feedback: string[] };
  performance_metrics: {
    predicted_score: number;
    actual_score: number;
    points_lost: number;
    percent_loss: number;
  };
  coach_recommendations: string[];
}

export class VirtualCoach {
  constructor(
    private bow: BowSetup,
    private arrow: ArrowSetup,
  ) {}

  /**
   * Full analysis pipeline:
   * 1. Equipment safety check
   * 2. Setup efficiency scoring
   * 3. Drag/skill loss calculation
   * 4. Synthesis & recommendations
   */
  analyzeSessionPerformance(
    shortScore: number,
    shortDist: number,
    shortFace: number,
    longScore: number,
    longDist: number,
    longFace: number,
  ): CoachAnalysis {
    const safetyWarnings = analyzeSetupSafety(this.bow, this.arrow);

    const discipline = longDist > 30 ? "outdoor" : "indoor";
    const setupAnalysis = scoreSetupEfficiency(this.bow, this.arrow, discipline as "indoor" | "outdoor");

    const dragAnalysis = calculateDragLoss(shortScore, shortDist, shortFace, longScore, longDist, longFace);

    const recommendations: string[] = [];
    if (dragAnalysis.percent_loss > 10.0) {
      recommendations.push("High Drag Loss detected (>10%).");
      if (setupAnalysis.gpp > 9.0) {
        recommendations.push("-> Your arrow is heavy (High GPP). Consider lighter points for 50m.");
      }
      if ((this.arrow.shaft_diameter_mm ?? 0) > 6.0) {
        recommendations.push("-> Your arrow is thick. Wind drift is likely the cause.");
      }
    }

    return {
      safety: safetyWarnings,
      setup_score: setupAnalysis,
      performance_metrics: dragAnalysis,
      coach_recommendations: recommendations,
    };
  }
}
